# Exploring Hacker News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

## Dataset

You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

## Aim of Project:

- Do "Ask HN" or "Show HN" post receive more comments on average?
- Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader

In [2]:
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

In [3]:
hn[0:4]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20']]

First row is the header, so we remove the particular row from the dataset and store it seperately.

In [4]:
headers = hn[0:1]

In [5]:
hn = hn[1:]

In [6]:
print(headers)
print(hn[0:4])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [7]:
ask_posts = []
show_posts = []
other_posts = []

In [8]:
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [9]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [10]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = row[4]
    total_ask_comments += int(num_comments)
avg_ask_comments = total_ask_comments / len(ask_posts)
print("Average Comments per ask post : ",avg_ask_comments)
print("Total comments on ask posts : ",total_ask_comments)

Average Comments per ask post :  14.038417431192661
Total comments on ask posts :  24483


In [11]:
total_show_comments = 0
for row in show_posts:
    num_comments = row[4]
    total_show_comments += int(num_comments)
avg_show_comments = total_show_comments / len(show_posts)
print("Average Comments per show post : ",avg_show_comments)
print("Total comments on show posts : ",total_show_comments)

Average Comments per show post :  10.31669535283993
Total comments on show posts :  11988


The "ask" posts have received more comments and the average comments per ask post is also higher than "show" posts.

In [12]:
import datetime as dt

In [13]:
result_list = []

In [14]:
for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

We create two dictionaries:

- counts_by_hour: contains the number of ask posts created during each hour of the day.
- comments_by_hour: contains the corresponding number of comments ask posts created at each hour received.

In [15]:
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

In [16]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

### We now calculate the average number of comments per post per hour.

In [17]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['06', 9.022727272727273],
 ['12', 9.41095890410959],
 ['18', 13.20183486238532],
 ['11', 11.051724137931034],
 ['09', 5.5777777777777775],
 ['20', 21.525],
 ['14', 13.233644859813085],
 ['10', 13.440677966101696],
 ['05', 10.08695652173913],
 ['02', 23.810344827586206],
 ['19', 10.8],
 ['22', 6.746478873239437],
 ['04', 7.170212765957447],
 ['08', 10.25],
 ['15', 38.5948275862069],
 ['16', 16.796296296296298],
 ['03', 7.796296296296297],
 ['00', 8.127272727272727],
 ['13', 14.741176470588234],
 ['23', 7.985294117647059],
 ['01', 11.383333333333333],
 ['21', 16.009174311926607],
 ['17', 11.46],
 ['07', 7.852941176470588]]

In [18]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

In [19]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

## Top 5 Hours for Ask Posts Comments

In [20]:
sorted_swap[0:5]

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21']]

In [21]:
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## Conclusion


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

## Further Analysis on Show Posts

In [22]:
ask_points = 0
for row in ask_posts:
    points = int(row[3])
    ask_points += points
    
ask_average_points = ask_points/len(ask_posts)

In [23]:
ask_average_points

15.061926605504587

In [24]:
show_points = 0
for row in show_posts:
    points = int(row[3])
    show_points += points
    
show_average_points = show_points/len(show_posts)

In [25]:
show_average_points

27.555077452667813

From the above values we can observe that although there are more show posts than ask posts, the average number of points on a show post are more higher than the average points on an ask post.

In [26]:
headers

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]

In [27]:
# Calculate the amount of show posts created during each hour of day and the number of points received.
import datetime as dt

result_list = []

for post in show_posts:
    result_list.append(
        [post[6], int(post[3])]
    )

posts_by_hour = {} #number of show posts hour-wise
points_by_hour = {} #number of points in the hour-wise
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    points = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in posts_by_hour:
        points_by_hour[time] += points
        posts_by_hour[time] += 1
    else:
        points_by_hour[time] = points
        posts_by_hour[time] = 1

### Average points per hour per post

In [28]:
avg_by_hour = []

for hr in points_by_hour:
    avg_by_hour.append([hr, points_by_hour[hr] / posts_by_hour[hr]])

In [29]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

In [30]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

In [31]:
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:10]:
    print(
        "{}: {:.2f} average points per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
23:00: 42.39 average points per post
12:00: 41.69 average points per post
22:00: 40.35 average points per post
00:00: 37.84 average points per post
18:00: 36.31 average points per post
11:00: 33.64 average points per post
19:00: 30.95 average points per post
20:00: 30.32 average points per post
15:00: 28.56 average points per post
16:00: 28.32 average points per post


Conclusion
The hour that receives the most points per post on average is 23:00, with an average of 42.39 points per post. There's nout much of a difference between the hours with the highest and second highest average number of points.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 23:00 as 11:00 pm est.

## Analysing Other Posts

In [32]:
other_comments = 0
for row in other_posts:
    comments = int(row[4])
    other_comments += comments
    
other_average_comments = other_comments/len(other_posts)

In [33]:
other_average_comments

26.8730371059672

As seen above the other posts receive a way higher number of comments than Ask Posts and Show Posts, as the "Other" category is a vary vast criteria, it would be difficult to point out which category in others receives the highest number of comments per post. 

In [34]:
other_points = 0
for row in other_posts:
    points = int(row[3])
    other_points += points
    
other_average_points = other_points/len(other_posts)

In [35]:
other_average_points

55.4067698034198

As seen above the other posts receive a way higher number of points than Ask Posts and Show Posts, as the "Other" category is a vary vast criteria, it would be difficult to point out which category in others receives the highest number of points per post.